lib/backend/k8s/resources/customresource.go

```diff
@@ -117,32 +115,60 @@ func (c *customK8sResourceClient) Update(kvp *model.KVPair) (*model.KVPair, erro
 	})
 	logContext.Debug("Update custom Kubernetes resource")
 
-	// Convert the KVPair to the K8s resource.
-	resIn, err := c.converter.FromKVPair(kvp)
-	if err != nil {
-		logContext.WithError(err).Info("Error updating resource")
-		return nil, err
+	// Create storage for the updated resource.
+	resOut := reflect.New(c.k8sResourceType).Interface().(CustomK8sResource)
+
+	providedRV := ""
+	if kvp.Revision != nil {
+		providedRV = kvp.Revision.(string)
 	}
 
-	// Send the update request using the name.
-	resOut := reflect.New(c.k8sResourceType).Interface().(CustomK8sResource)
-	name := resIn.GetObjectMeta().GetName()
-	logContext = logContext.WithField("Name", name)
-	logContext.Debug("Update resource by name")
-	err = c.restClient.Put().
-		Resource(c.resource).
-		Namespace("kube-system").
-		Body(resIn).
-		Name(name).
-		Do().Into(resOut)
-	if err != nil {
-		logContext.WithError(err).Info("Error updating resource")
-		return nil, K8sErrorToCalico(err, kvp.Key)
+	var updateError error
+	for i := 0; i < 5; i++ {
+		// If no revision was passed, get the object to use its latest Revision number.
+		// If a revision was passed, then we should just use that.
+		if providedRV == "" {
+			logContext.Debug("Querying for resource version")
+			k, err := c.Get(kvp.Key)
+			if err != nil {
+				return nil, err
+			}
+
+			if k.Revision != nil {
+				kvp.Revision = k.Revision.(string)
+				logContext.Debugf("Set resource version to %s", kvp.Revision)
+			}
+		}
+
+		// Convert the KVPair to a K8s resource.
+		resIn, err := c.converter.FromKVPair(kvp)
+		if err != nil {
+			logContext.WithError(err).Info("Error updating resource")
+			return nil, err
+		}
+
+		// Send the update request using the name.
+		name := resIn.GetObjectMeta().GetName()
+		logContext = logContext.WithField("Name", name)
+		logContext.Debug("Update resource by name")
+		updateError = c.restClient.Put().
+			Resource(c.resource).
+			Body(resIn).
+			Name(name).
+			Do().Into(resOut)
+		if updateError == nil {
```

Comment:
This logic is a bit different from what was here before.

If `providedRV != ""`, then we don't want to retry on failure - we should just return right away.

We should only retry when:
- No Revision was provided on the KVP, AND
- The error we got back was due to a CAS error.